pkg/dogstatsd/parse_metrics.go

```diff
@@ -0,0 +1,142 @@
+package dogstatsd
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+)
+
+type metricType int
+
+const (
+	gaugeType metricType = iota
+	countType
+	distributionType
+	histogramType
+	setType
+	timingType
+)
+
+var (
+	gaugeSymbol        = []byte("g")
+	countSymbol        = []byte("c")
+	histogramSymbol    = []byte("h")
+	distributionSymbol = []byte("d")
+	setSymbol          = []byte("s")
+	timingSymbol       = []byte("ms")
+
+	tagsFieldPrefix       = []byte("#")
+	sampleRateFieldPrefix = []byte("@")
+)
+
+type dogstatsdMetricSample struct {
+	name       string
+	value      float64
+	setValue   string
+	metricType metricType
+	sampleRate float64
+	tags       []string
+}
+
+// sanity checks a given message against the metric sample format
+func hasMetricSampleFormat(message []byte) bool {
+	if message == nil {
+		return false
+	}
+	separatorCount := bytes.Count(message, fieldSeparator)
+	if separatorCount < 1 || separatorCount > 3 {
+		return false
+	}
+	return true
+}
+
+func parseMetricSampleNameAndRawValue(rawNameAndValue []byte) ([]byte, []byte, error) {
+	sepIndex := bytes.Index(rawNameAndValue, colonSeparator)
+	if sepIndex == -1 {
+		return nil, nil, fmt.Errorf("invalid name and value: %q", rawNameAndValue)
+	}
+	rawName := rawNameAndValue[:sepIndex]
+	rawValue := rawNameAndValue[sepIndex+1:]
+	if len(rawName) == 0 || len(rawValue) == 0 {
+		return nil, nil, fmt.Errorf("invalid name and value: %q", rawNameAndValue)
+	}
+	return rawName, rawValue, nil
+}
+
+func parseMetricSampleMetricType(rawMetricType []byte) (metricType, error) {
+	switch {
+	case bytes.Equal(rawMetricType, gaugeSymbol):
+		return gaugeType, nil
+	case bytes.Equal(rawMetricType, countSymbol):
+		return countType, nil
+	case bytes.Equal(rawMetricType, histogramSymbol):
+		return histogramType, nil
+	case bytes.Equal(rawMetricType, distributionSymbol):
+		return distributionType, nil
+	case bytes.Equal(rawMetricType, setSymbol):
+		return setType, nil
+	case bytes.Equal(rawMetricType, timingSymbol):
+		return timingType, nil
+	}
+	return 0, fmt.Errorf("invalid metric type: %q", rawMetricType)
+}
+
+func parseMetricSampleSampleRate(rawSampleRate []byte) (float64, error) {
+	return strconv.ParseFloat(string(rawSampleRate), 64)
+}
+
+func parseMetricSample(message []byte) (dogstatsdMetricSample, error) {
+	// fast path to eliminate most of the gibberish
+	// especially important here since all the unidentified garbage gets
+	// identified as metrics
+	if !hasMetricSampleFormat(message) {
+		return dogstatsdMetricSample{}, fmt.Errorf("invalid dogstatsd message format: %q", message)
+	}
```

Comment:
What's the frequency of gibberish input? That shouldn't happen in normal situation, isn't it?

Asking this because if the frequency is not high, the verifications `hasMetricSampleFormat` (same for service checks, events) might not worth it since we are catching those erroneous cases later anyway.

Reply:
Internally we don't have a lot of parsing errors but we can't really know for our clients. This check was ported from the old parsing logic to try to limit the risk that we might start parsing things that we did not before.

Reply:
👋We do have a lot of gibberish: SQL queries, HTML, CSS, and JS is something we regularly see hitting our metrics pre-filtering endpoint that cleans that out before sending to the datadog-agent. Most of those are just due to various bugs that developers don't know about.

(Disclaimer: speaking as a customer, running _lots_ of hosts.)